File: app/src/main/java/com/example/solar_alarm/Data/Repositories/SolarAlarmRepository.java
package com.example.solar_alarm.Data.Repositories;

import android.os.Build;

import androidx.annotation.RequiresApi;
import androidx.lifecycle.LiveData;

import com.example.solar_alarm.Data.Daos.SolarAlarmDao;
import com.example.solar_alarm.Data.SolarAlarmDatabase;
import com.example.solar_alarm.Data.Tables.SolarAlarm;

import java.util.List;

@RequiresApi(api = Build.VERSION_CODES.O)
public class SolarAlarmRepository extends RepositoryBase
{
    private final SolarAlarmDao solarAlarmDao;
    private final LiveData<List<SolarAlarm>> solarAlarmLiveData;

    @RequiresApi(api = Build.VERSION_CODES.O)
    public SolarAlarmRepository()
    {
        solarAlarmDao = _SolarAlarmDatabase.solarAlarmDao();
        solarAlarmLiveData = solarAlarmDao.getAll();
    }

    @RequiresApi(api = Build.VERSION_CODES.O)
    public void Insert(SolarAlarm solarAlarm)
    {
        SolarAlarmDatabase.databaseWriteExecutor.execute(() -> solarAlarmDao.Insert(solarAlarm));
    }

    @RequiresApi(api = Build.VERSION_CODES.O)
    public void Update(SolarAlarm solarAlarm)
    {
        SolarAlarmDatabase.databaseWriteExecutor.execute(() -> solarAlarmDao.Update(solarAlarm));
    }

    @RequiresApi(api = Build.VERSION_CODES.O)
    public void Delete(SolarAlarm solarAlarm)
    {
        SolarAlarmDatabase.databaseWriteExecutor.execute(() -> solarAlarmDao.delete(solarAlarm));
    }

    public boolean isSolarAlarmNameLocationIDExists(String name, int locationId)
    {
        return solarAlarmDao.isSolarAlarmNameLocationIDPairExists(name, locationId);
    }
    public LiveData<List<SolarAlarm>> getAll() {return solarAlarmLiveData;}
}

File: app/src/main/java/com/example/solar_alarm/Data/Repositories/SolarAlarmRepository.kt
package com.example.solar_alarm.data.repositories

import android.os.Build
import androidx.annotation.RequiresApi
import androidx.annotation.WorkerThread
import com.example.solar_alarm.data.daos.SolarAlarmDao
import com.example.solar_alarm.data.tables.SolarAlarm
import com.example.solar_alarm.data.tables.SolarAlarmWithDetails
import kotlinx.coroutines.flow.Flow
import javax.inject.Inject
import javax.inject.Singleton

@RequiresApi(api = Build.VERSION_CODES.O)
@Singleton
class SolarAlarmRepository @Inject constructor(private val solarAlarmDao: SolarAlarmDao)
{
    val all: Flow<List<SolarAlarm>> = solarAlarmDao.getAll()
    val allWithDetails: Flow<List<SolarAlarmWithDetails>> = solarAlarmDao.getAllWithDetails()

    @Suppress("RedundantSuspendModifier")
    @WorkerThread
    suspend fun insert(solarAlarm: SolarAlarm)
    {
        solarAlarmDao.insert(solarAlarm)
    }

    @WorkerThread
    suspend fun getById(id: Int): SolarAlarm?
    {
        return solarAlarmDao.getById(id)
    }

    @WorkerThread
    suspend fun update(solarAlarm: SolarAlarm)
    {
        solarAlarmDao.update(solarAlarm)
    }

    @WorkerThread
    suspend fun delete(solarAlarm: SolarAlarm)
    {
        solarAlarmDao.delete(solarAlarm)
    }

    @WorkerThread
    suspend fun isSolarAlarmNameLocationIdPairExists(solarAlarm: SolarAlarm): Boolean
    {
        return solarAlarmDao.isSolarAlarmNameLocationIdPairExists(solarAlarm.Name, solarAlarm.LocationId)
    }
}
